### PR TITLE
Remove downlevel-dts and legacy type support

### PR DESCRIPTION
**Pull request by alex-page**
### WHY are these changes introduced?

Polaris currently supports multiple TypeScript versions by building our types twice. This breaking change in v11 removes this functionality and we will only support the current version in the library.

This used to be required to build the props table on legacy.polaris.shopify.com. This is no longer a requirement with @martenbjork brilliant work.

### WHAT is this pull request doing?

Removes legacy type support for the library.

### CODE REVIEW

File: polaris-react/scripts/build-validate.js [hunk 1]
```diff
@@ -81,12 +80,6 @@ function validateEsNextBuild() {
   assert.ok(jsContent.includes('"hidden": "Polaris-Avatar--hidden_riqie"'));
 }
 
-function validateAncillaryOutput() {
-  assert.ok(fs.existsSync('./build/ts/latest/src/index.d.ts'));
```
Comment (BPScott) [inline polaris-react/scripts/build-validate.js:85]: Asserting that the types are present in the build still seems like a useful thing to test. This PR removes the `build/ts/3.4` directory, but the main types remain.

File: yarn.lock [hunk 2]
```diff
@@ -21687,16 +21679,16 @@ typedarray@^0.0.6:
...
-  version "4.6.3"
-  resolved "https://registry.yarnpkg.com/typescript/-/typescript-4.6.3.tgz#eefeafa6afdd31d725584c67a0eaba80f6fc6c6c"
-  integrity sha512-yNIatDa5iaofVozS/uQJEl3JRWLKKGJKh6Yaiv0GLGSuhpFJe7P3SbHZ8/yjAHRQwKRoA6YZqlfjXWmVzoVSMw==
-
 typescript@^4.3.2:
   version "4.8.4"
   resolved "https://registry.yarnpkg.com/typescript/-/typescript-4.8.4.tgz#c464abca159669597be5f96b8943500b238e60e6"
   integrity sha512-QCh+85mCy+h0IGff8r5XWzOVSbBO+KfeYrMQh7NJ58QujwcE22u+NUSmUxqF+un70P9GXKxa2HCNiTTMJknyjQ==
 
+typescript@^4.6.3:
```
Comment (BPScott) [inline yarn.lock]: The typescript dependency should be deduped.

**Review by BPScott — COMMENTED** [review]
You can also delete the typesVersions config from the package.json too - https://github.com/Shopify/polaris/blob/main/polaris-react/package.json#L40-L46


Is this an opportunity to also adjust the type output location? Currently they're in `build/ts/latest` (per https://github.com/Shopify/polaris/blob/main/polaris-react/tsconfig.json#L7 (which says where TS should create them) and https://github.com/Shopify/polaris/blob/main/polaris-react/package.json#L39 (which says where consumers can find them). perhaps that can be shortened to `build/ts` now?

File: polaris-react/src/components/OptionList/tests/OptionList.test.tsx [hunk 3]
```diff
@@ -539,12 +539,9 @@ function firstOption(
...
   sections?: OptionListProps['sections'],
 ): string {
-  const firstOptionsValue =
-    options == null || options === [] ? '' : options[0].value;
+  const firstOptionsValue = options == null ? '' : options[0].value;
   const firstSectionOptionsValue =
-    sections == null || sections === [] || sections[0].options === []
-      ? ''
-      : sections[0].options[0].value;
+    sections == null ? '' : sections[0].options[0].value;
```
Comment (alex-page) [inline polaris-react/src/components/OptionList/tests/OptionList.test.tsx:544]: The array checks were unnecessary with the null check. Removed them and tests function the same.

File: polaris-react/src/utilities/merge.ts [hunk 4]
```diff
@@ -24,9 +24,13 @@ export function merge<TSource1, TSource2, TSource3, TSource4, TSource5>(
...
-export function merge<TSource1, TSource2, TSource3, TSource4, TSource5>(
-  ...objs: (TSource1 | TSource2 | TSource3 | TSource4 | TSource5)[]
-) {
+export function merge<
+  TSource1 extends GeneralObject,
+  TSource2 extends GeneralObject,
+  TSource3 extends GeneralObject,
+  TSource4 extends GeneralObject,
+  TSource5 extends GeneralObject,
+>(...objs: (TSource1 | TSource2 | TSource3 | TSource4 | TSource5)[]) {
```
Comment (alex-page) [inline polaris-react/src/utilities/merge.ts:33]: TypeScript wanted these types to `extend GeneralObject` so I just spammed it in here 🤷

**Review by aaronccasanova — APPROVED** [review]

**Pull request merged**